[2024-09-13 Fiddler](https://thefiddler.substack.com/p/can-you-shut-the-box)
====================

When there is one unflipped tile, the probability of winning is $1/6$.

When there are two unflipped tiles, if the tiles 1 and 2, the probability of winning is
$1/6+2/6\cdot1/6 = 2/9$, the probability of rolling 3 plus the probability of rolling 1 then 2
plus the probability of rolling 2 then 1.

If the unflipped tiles are 1 and 3, 1 and 4, 1 and 5, 2 and 3, 2 and 4, the probability of winning
is similarly $2/9$.

If the unflipped tiles are 2 and 5, 2 and 6, 3 and 4, 3 and 5, 3 and 6, 4 and 5, 4 and 6, 5 and 6,
the probability of winning is $1/18$.

If the unflipped tiles are 123, the probability of winning is $3/6\cdot2/9+2/6\cdot1/6+1/6 = 1/3$,
because when rolling a 3, it's better to flip the 3 than the 1 and 2.

And so on...

In [1]:
p_1 = 1/6
p_2 = 1/6
p_3 = 1/6
p_4 = 1/6
p_5 = 1/6
p_6 = 1/6

p_12 = 1/6*p_2 + 1/6*p_1 + 1/6
p_13 = 1/6*p_3 + 1/6*p_1 + 1/6
p_14 = 1/6*p_4 + 1/6*p_1 + 1/6
p_15 = 1/6*p_5 + 1/6*p_1 + 1/6
p_16 = 1/6*p_6 + 1/6*p_1
p_23 = 1/6*p_3 + 1/6*p_2 + 1/6
p_24 = 1/6*p_4 + 1/6*p_2 + 1/6
p_25 = 1/6*p_5 + 1/6*p_2
p_26 = 1/6*p_6 + 1/6*p_2
p_34 = 1/6*p_4 + 1/6*p_3
p_35 = 1/6*p_5 + 1/6*p_3
p_36 = 1/6*p_6 + 1/6*p_3
p_45 = 1/6*p_5 + 1/6*p_5
p_46 = 1/6*p_6 + 1/6*p_4
p_56 = 1/6*p_6 + 1/6*p_5

p_123 = 1/6*p_23 + 1/6*p_13 + 1/6*max(p_3,p_12) + 1/6*p_2 + 1/6*p_1 + 1/6
p_124 = 1/6*p_24 + 1/6*p_14 + 1/6*p_4 + 1/6*p_12 + 1/6*p_2 + 1/6*p_1
p_125 = 1/6*p_25 + 1/6*p_15 + 1/6*p_5 + 1/6*p_12 + 1/6*p_2
p_126 = 1/6*p_26 + 1/6*p_16 + 1/6*p_6 + 1/6*p_12
p_134 = 1/6*p_34 + 1/6*p_14 + 1/6*max(p_4,p_13) + 1/6*p_3
p_135 = 1/6*p_35 + 1/6*p_15 + 1/6*p_5 + 1/6*p_13 + 1/6*p_3
p_136 = 1/6*p_36 + 1/6*p_16 + 1/6*p_6 + 1/6*p_13
p_145 = 1/6*p_45 + 1/6*p_15 + 1/6*max(p_5,p_14) + 1/6*p_4
p_146 = 1/6*p_46 + 1/6*p_16 + 1/6*p_6 + 1/6*p_14
p_156 = 1/6*p_56 + 1/6*p_16 + 1/6*max(p_6,p_15)
p_234 = 1/6*p_34 + 1/6*p_24 + 1/6*p_23 + 1/6*p_4 + 1/6*p_3
p_235 = 1/6*p_35 + 1/6*p_25 + 1/6*max(p_5,p_23)
p_236 = 1/6*p_36 + 1/6*p_26 + 1/6*p_6 + 1/6*p_23
p_245 = 1/6*p_45 + 1/6*p_25 + 1/6*p_24 + 1/6*p_5
p_246 = 1/6*p_46 + 1/6*p_26 + 1/6*max(p_6,p_24)
p_256 = 1/6*p_56 + 1/6*p_26 + 1/6*p_25
p_345 = 1/6*p_45 + 1/6*p_35 + 1/6*p_34
p_346 = 1/6*p_46 + 1/6*p_36 + 1/6*p_34
p_356 = 1/6*p_56 + 1/6*p_36 + 1/6*p_35
p_456 = 1/6*p_56 + 1/6*p_46 + 1/6*p_45

p_1234 = 1/6*p_234 + 1/6*p_134 + 1/6*max(p_34,p_124) + 1/6*max(p_24,p_123) + 1/6*max(p_23,p_14) + 1/6*max(p_4,p_13)
p_1235 = 1/6*p_235 + 1/6*p_135 + 1/6*max(p_35,p_125) + 1/6*p_25 + 1/6*max(p_15,p_123) + 1/6*max(p_5,p_23)
p_1236 = 1/6*p_236 + 1/6*p_136 + 1/6*max(p_36,p_126) + 1/6*p_26 + 1/6*p_16 + 1/6*max(p_6,p_123)
p_1245 = 1/6*p_245 + 1/6*p_145 + 1/6*p_45 + 1/6*p_125 + 1/6*max(p_25,p_124) + 1/6*max(p_24,p_15)
p_1246 = 1/6*p_246 + 1/6*p_146 + 1/6*p_46 + 1/6*p_126 + 1/6*p_26 + 1/6*max(p_16,p_124)
p_1256 = 1/6*p_256 + 1/6*p_156 + 1/6*p_56 + 1/6*p_126 + 1/6*max(p_26,p_125)
p_1345 = 1/6*p_345 + 1/6*p_145 + 1/6*max(p_45,p_135) + 1/6*max(p_35,p_134) + 1/6*p_34
p_1346 = 1/6*p_346 + 1/6*p_146 + 1/6*max(p_46,p_136) + 1/6*p_36 + 1/6*p_134
p_1356 = 1/6*p_356 + 1/6*p_156 + 1/6*p_56 + 1/6*p_136 + 1/6*max(p_36,p_135)
p_1456 = 1/6*p_456 + 1/6*p_156 + 1/6*max(p_56,p_146) + 1/6*max(p_46,p_145)
p_2345 = 1/6*p_345 + 1/6*p_245 + 1/6*p_235 + 1/6*max(p_45,p_234) + 1/6*p_35
p_2346 = 1/6*p_346 + 1/6*p_246 + 1/6*p_236 + 1/6*p_46 + 1/6*max(p_36,p_234)
p_2356 = 1/6*p_356 + 1/6*p_256 + 1/6*max(p_56,p_236) + 1/6*p_235
p_2456 = 1/6*p_456 + 1/6*p_256 + 1/6*p_246 + 1/6*max(p_56,p_245)
p_3456 = 1/6*p_456 + 1/6*p_356 + 1/6*p_346 + 1/6*p_345

p_12345 = 1/6*p_2345 + 1/6*p_1345 + 1/6*max(p_345,p_1245) + 1/6*max(p_245,p_1235) + 1/6*max(p_235,p_145,p_1234) + 1/6*max(p_45,p_234,p_135)
p_12346 = 1/6*p_2346 + 1/6*p_1346 + 1/6*max(p_346,p_1246) + 1/6*max(p_246,p_1236) + 1/6*max(p_236,p_146) + 1/6*max(p_46,p_136,p_1234)
p_12356 = 1/6*p_2356 + 1/6*p_1356 + 1/6*max(p_356,p_1256) + 1/6*p_256 + 1/6*max(p_156,p_1236) + 1/6*max(p_56,p_236,p_1235)
p_12456 = 1/6*p_2456 + 1/6*p_1456 + 1/6*p_456 + 1/6*p_1256 + 1/6*max(p_256,p_1246) + 1/6*max(p_246,p_156,p_1245)
p_13456 = 1/6*p_3456 + 1/6*p_1456 + 1/6*max(p_456,p_1356) + 1/6*max(p_356,p_1346) + 1/6*max(p_346,p_1345)
p_23456 = 1/6*p_3456 + 1/6*p_2456 + 1/6*p_2356 + 1/6*max(p_456,p_2346) + 1/6*max(p_356,p_2345)

p_123456 = 1/6*p_23456 + 1/6*p_13456 + 1/6*max(p_3456,p_12456) + 1/6*max(p_2456,p_12356) + 1/6*max(p_2356,p_1456,p_12346) + 1/6*max(p_456,p_2346,p_1356,p_12345)
(p_123456,numerical_approx(p_123456))

(145/1944, 0.0745884773662551)

Extra credit
------------

First, define the probability of rolling a given total

In [2]:
roll(n) = (6-abs(n-7))/36

Next, get the possible remaining tiles for a given sum

In [3]:
def remaining_tiles(tiles, sum):
    result = set()
    for tile in tiles:
        if tile == sum:
            result |= {frozenset(tiles - {tile})}
        elif tile < sum:
            result |= remaining_tiles(frozenset(tiles - {tile}), sum-tile)
    return result

Finally, get the probability of winning

In [4]:
from functools import cache

@cache
def p6(tiles):
    if len(tiles) == 0:
        return 1
    result = 0
    for sum in [1..6]:
        result += 1/6*max([p6(next_tiles) for next_tiles in remaining_tiles(tiles,sum)] + [0])
    return result

@cache
def p12(tiles):
    if len(tiles) == 0:
        return 1
    result = 0
    for sum in [2..12]:
        result += roll(sum)*max([p12(next_tiles) for next_tiles in remaining_tiles(tiles,sum)] + [0])
    return result

Check the one die result

In [5]:
p6(frozenset([1..6])), p_123456

(145/1944, 145/1944)

In [6]:
p_123456789 = p12(frozenset([1..9]))
(p_123456789, numerical_approx(p_123456789))

(466473281/6530347008, 0.0714316223056060)

Simulations
-----------

Running [simulations](20240913.go).

I'll use the above calculations to decide which combination of tiles to flip in the simulations.

First, check the code against the above calculations.

In [7]:
show(numerical_approx(p6(frozenset([1..6])),digits=5))
show([numerical_approx(p6(frozenset(filter(lambda x: x != i, [1..6]))),digits=5) for i in [1..6]])
show(numerical_approx(p12(frozenset([1..9])),digits=5))
show([numerical_approx(p12(frozenset(filter(lambda x: x != i, [1..9]))),digits=5) for i in [1..9]])

0.074588

[0.033951, 0.043210, 0.064815, 0.075617, 0.10185, 0.12809]

0.071432

[0.042214,
 0.045190,
 0.051068,
 0.055327,
 0.061085,
 0.064993,
 0.071253,
 0.084399,
 0.10150]

The code passes the checks.

    $ go run 20240913.go check
    0.074588
     0.033951 0.043210 0.064815 0.075617 0.101852 0.128086
    0.071432
     0.042214 0.045190 0.051068 0.055327 0.061085 0.064993 0.071253 0.084399 0.101502

And the simulation results agree.

    $ go run 20240913.go
    f: 0.074662                                                                     
    ec: 0.071431

Making the rounds
-----------------

[Simulations](20240913mtr.go) suggest a uniform distribution between 0 and 1.

Let $q = (xy)^z$.  Given $x$, $y$, and $q$, $z$ is

In [8]:
%display latex
q,x,y,z = var('q,x,y,z')
solve(q == (x*y)^z, z)

[z == log(q)/log(x*y)]

The probability that $x$ is between $x$ and $x+dx$, $y$ is between $y$ and $y+dy$, and $z$
is between $z$ and $z+dz$, where $0<x<1$, $0<y<1$, $0<z<1$, $dx$, $dy$, and $dz$ are small,
is $dx\,dy\,dz$.

The probability that $q$ is between $q$ and $q+dq$, where $0<q<1$ and $dq$ is small, would
be $dq$ if the distribution is uniform.  In any case, it's the sum of all the contributions
from $x$, $y$, and $z$, where $q < (xy)^z < q+dq$.  So, given $q$, it's where
$-\log(xy) > -\log q$, since $0<z<1$, so $0 < y < q/x$,

$$ P(q)\,dq = \int_0^1 dx\,\int_0^{q/x} dy\,dz $$

And $dz = \frac{dq}{q\log(xy)}$, so

$$ P(q) = \int_0^1 dx\,\int_0^{q/x} dy\,\frac{1}{q\log(xy)} $$

That integral is divergent, so there's something wrong with my reasoning.

In [9]:
assume(q > 0)
assume(q < 1)
assume(x > 0)
assume(x < 1)
integral(1/(q*log(x*y)),y,0,q/x)

Ei(log(q))/(q*x)